spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingEnhancer.java

```diff
@@ -0,0 +1,208 @@
+		}
+	}
+
+	/**
+	 * Because the parser can find the alias of the FROM clause, there is no need to "find it" in advance.
+	 *
+	 * @param sort the sort specification to apply.
+	 * @param alias IGNORED
+	 * @return
+	 */
+	@Override
+	public String applySorting(Sort sort, String alias) {
+		return applySorting(sort);
+	}
+
+	/**
+	 * Resolves the alias for the entity in the FROM clause from the JPA query. Since the {@link QueryParser} can already
+	 * find the alias when generating sorted and count queries, this is mainly to serve test cases.
+	 */
+	@Override
+	public String detectAlias() {
+
+		try {
+			ParserRuleContext parsedQuery = queryParser.parse();
+
+			if (parsedQuery == null) {
+
+				LOG.warn("Failed to parse " + queryParser.getQuery() + ". See console for more details.");
+				return null;
+			}
+
+			return queryParser.findAlias(parsedQuery);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.warn(e);
+			return null;
+		}
+	}
+
+	/**
+	 * Creates a count query from the original query, with no count projection.
+	 * 
+	 * @return Guaranteed to be not {@literal null};
+	 */
+	@Override
+	public String createCountQueryFor() {
+		return createCountQueryFor(null);
+	}
+
+	/**
+	 * Create a count query from the original query, with potential custom projection.
+	 *
+	 * @param countProjection may be {@literal null}.
+	 */
+	@Override
+	public String createCountQueryFor(@Nullable String countProjection) {
+
+		try {
+			ParserRuleContext parsedQuery = queryParser.parse();
+
+			if (parsedQuery == null) {
+				return "";
+			}
+
+			return queryParser.createCountQuery(parsedQuery, countProjection);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.warn(e);
+			throw new IllegalArgumentException(e);
+		}
+	}
+
+	/**
+	 * Checks if the select clause has a new constructor instantiation in the JPA query.
+	 *
+	 * @return Guaranteed to return {@literal true} or {@literal false}.
+	 */
+	@Override
+	public boolean hasConstructorExpression() {
+
+		try {
+			ParserRuleContext parsedQuery = queryParser.parse();
+
+			if (parsedQuery == null) {
+				return false;
+			}
+
+			return queryParser.hasConstructor(parsedQuery);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.warn(e);
+			return false;
+		}
+	}
+
+	/**
+	 * Looks up the projection of the JPA query. Since the {@link QueryParser} can already find the projection when
+	 * generating sorted and count queries, this is mainly to serve test cases.
+	 */
+	@Override
+	public String getProjection() {
+
+		try {
+			ParserRuleContext parsedQuery = queryParser.parse();
+
+			if (parsedQuery == null) {
+				return "";
+			}
+
+			return queryParser.projection(parsedQuery);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.debug(e);
+			return "";
+		}
+	}
+
+	/**
+	 * Since the {@link QueryParser} can already fully transform sorted and count queries by itself, this is a placeholder
+	 * method.
+	 *
+	 * @return empty set
+	 */
+	@Override
+	public Set<String> getJoinAliases() {
```

Comment:
`getJoinAliases` seems to be not used in production code. We should deprecate it and make it a `default` method in the interface.